### PR TITLE
Use Transformation to get rid of the ctx parameter in generators

### PR DESCRIPTION
**Pull request by vil1**

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/Result.scala [hunk 1]
```diff
@@ -47,6 +48,30 @@ trait TransformationConstructors[S] {
   def set(newState: S): Transformation[S, Unit] = new Transformation(OkResult(_ => OkResult((newState, ()))))
   def update[T](f: PartialFunction[S, S]): Transformation[S, Unit] = new Transformation(
     OkResult(s => OkResult((f.applyOrElse(s, identity[S]), ()))))
+
+  def updateGenCtx(f: GeneratorContext => GeneratorContext): Transformation[Phase, Unit] = new Transformation(OkResult {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/Result.scala:52]: these methods belong to `CodeGenerator` trait.

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/ExpressionGenerator.scala [hunk 2]
```diff
@@ -11,63 +11,64 @@ class ExpressionGenerator extends BaseSQLGenerator[ir.Expression] with Transform
...
+      case ua: ir.UpdateAction => updateAction(ua)
+      case a: ir.Assign => assign(a)
+      case opts: ir.Options => options(opts)
+      case i: ir.KnownInterval => interval(i)
+      case s: ir.ScalarSubquery => scalarSubquery(s)
+      case c: ir.Case => caseWhen(c)
+      case w: ir.Window => window(w)
+      case o: ir.SortOrder => sortOrder(o)
+      case ir.Exists(subquery) =>
+        withGenCtx.flatMap(ctx => code"EXISTS (${ctx.logical.generate(subquery)})")
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/ExpressionGenerator.scala:57]: ok, if this keeps working - i'm fine. btw, what do you think about this sugar:
```suggestion
      case ir.Exists(subquery) => withGeneratorContext { ctx =>
        code"EXISTS (${ctx.logical.generate(subquery)})"
      }
```

`withGeneratorContext` to be on `CodeGenerator` trait

File: core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala [hunk 3]
```diff
@@ -30,7 +25,7 @@ class SnowflakePlanParser extends PlanParser[SnowflakeParser] {
       new SnowflakeCallMapper,
       ir.AlwaysUpperNameForCallFunction,
       new UpdateToMerge,
-      new CastParseJsonToFromJson(generator),
+      new CastParseJsonToFromJson,
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakePlanParser.scala:28]: nice

**Review by nfx — CHANGES_REQUESTED** [review]
please move changes from `TransformationConstructors` to `CodeGenerator` trait

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/ExpressionGenerator.scala (hunk 2 above)
Comment (vil1) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/ExpressionGenerator.scala]: You're right, that's more convenient. Done.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**